lib/std/date/gregorian.zig

```diff
@@ -0,0 +1,397 @@
+
+            const N = @as(UEpochDays, @bitCast(days)) +% K;
+
+            const a1 = 4;
+            const b1 = 3;
+            const N_1 = a1 * N + b1;
+            const C = N_1 / era_days;
+            const N_C: UIntFitting(36_564) = div(mod(N_1, era_days), a1);
+
+            const N_2 = a1 * @as(UIntFitting(146_099), N_C) + b1;
+            // n % 1461 == 2939745 * n % 2^32 / 2939745,
+            // for all n in [0, 28825529)
+            assert(N_2 < 28_825_529);
+            const a2 = 2_939_745;
+            const b2 = 0;
+            const P_2_max = 429493804755;
+            const P_2 = a2 * @as(UIntFitting(P_2_max), N_2) + b2;
+            const Z: UIntFitting(99) = div(P_2, (1 << 32));
+            const N_Y: UIntFitting(365) = div(mod(P_2, (1 << 32)), a2 * a1);
+
+            // (5 * n + 461) / 153 == (2141 * n + 197913) /2^16,
+            // for all n in [0, 734)
+            assert(N_Y < 734);
+            const a3 = 2_141;
+            const b3 = 197_913;
+            const N_3 = a3 * @as(UIntFitting(979_378), N_Y) + b3;
+
+            return (Computational{
+                .year = 100 * C + Z,
+                .month = div(N_3, 1 << 16),
+                .day = div(mod(N_3, (1 << 16)), a3),
+            }).toGregorian(N_Y);
+        }
+
+        pub fn toEpoch(self: Self) EpochDays {
+            // This function is Figure 13 of the paper.
+            const c = Computational.fromGregorian(self);
+            const C = c.year / 100;
+
+            const y_star = days_in_year.numerator * c.year / 4 - C + C / 4;
+            const days_in_5mo = 31 + 30 + 31 + 30 + 31;
+            const m_star = (days_in_5mo * @as(UEpochDays, c.month) - 457) / 5;
+            const N = y_star + m_star + c.day;
+
+            return @as(EpochDays, @intCast(N)) - K;
+        }
+
+        pub const MonthAdd = std.meta.Int(.signed, @typeInfo(IEpochDays).Int.bits - std.math.log2_int(u16, 12));
+
+        pub fn add(self: Self, year: Year, month: MonthAdd, day: IEpochDays) Self {
+            const m = month + self.month.numeric() - 1;
+            const y = self.year + year + @divFloor(m, 12);
+
+            const ym_epoch_day = Self{
+                .year = @intCast(y),
+                .month = @enumFromInt(std.math.comptimeMod(m, 12) + 1),
+                .day = 1,
+            };
+
+            var epoch_days = ym_epoch_day.toEpoch();
+            epoch_days += day + self.day - 1;
+
+            return fromEpoch(epoch_days);
+        }
+
+        pub const Weekday = WeekdayT;
+        pub fn weekday(self: Self) Weekday {
+            // 1970-01-01 is a Thursday.
+            const epoch_days = self.toEpoch() +% Weekday.thu.numeric();
+            return @enumFromInt(std.math.comptimeMod(epoch_days, 7));
+        }
+    };
+}
+
+/// Epoch is days since 1970
+pub fn Date(comptime Year: type, epoch: comptime_int) type {
+    const shift = solveShift(Year, epoch) catch unreachable;
+    return DateAdvanced(Year, epoch, shift);
+}
```

Comment:
One more thing that came to my mind; I see a lot of `comptime` going on (date fields, time fields) - wouldn't this be an issue if some Date(Time) variable should be updated at runtime?

Reply:
The only comptime thing that doesn't need to be is `Date`'s `epoch` and `shift`. This allows verifying at comptime that the full range of `Year` is supported by EAFs so that `fn toEpoch()` need not check the that the value of `year` is in range.

We could allow arbitrary runtime epochs by moving this to a runtime assert, but that breaks the `zig zen` of:
> * Compile errors are better than runtime crashes.

...and I'm not sure how useful it'd be. I believe there are an enumerable number of `epoch`s.

Reply:
ok, seems I misread that. In `Date(comptime Year: type, epoch: comptime_int)`, "Year" is actually a type. That confused me. "YearType" etc. might be more intuitive (you're not saying "EpochD" either, for "EpochDays"). At least for me ;-)

A user-defined epoch should still be a `const` IIUC; an epoch should change suddenly.